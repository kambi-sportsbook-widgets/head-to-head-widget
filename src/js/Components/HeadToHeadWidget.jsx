import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { widgetModule, translationModule } from 'kambi-widget-core-library'
import Header from './Header/Header'
import EventInfo from './EventInfo/EventInfo'
import BoxContainer from './BoxContainer/BoxContainer'

class HeadToHeadWidget extends Component {
  /**
   * Called after component mounts
   */
  componentDidMount() {
    widgetModule.adaptWidgetHeight()
  }
  /**
   * Called after updating component's DOM.
   */
  componentDidUpdate() {
    widgetModule.adaptWidgetHeight()
  }

  render() {
    const t = translationModule.getTranslation.bind(translationModule)
    return (
      <div>
        <Header title={t('Head to head')} />
        {this.props.events.map(event => {
          return event.lastEvents.map((lastEvent, i) => (
            <BoxContainer key={i}>
              <EventInfo
                homeName={lastEvent.homeParticipant}
                homeScore={lastEvent.homeScore}
                awayName={lastEvent.awayParticipant}
                awayScore={lastEvent.awayScore}
                start={lastEvent.start}
              />
            </BoxContainer>
          ))
        })}
      </div>
    )
  }
}

HeadToHeadWidget.propTypes = {
  events: PropTypes.arrayOf(PropTypes.object).isRequired,
}

export default HeadToHeadWidget
