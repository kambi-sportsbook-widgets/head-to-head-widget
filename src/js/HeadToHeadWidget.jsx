import React from 'react';
import { coreLibrary, widgetModule, translationModule, statisticsModule } from 'widget-core-library';
import ScoreBar from './ScoreBar';

/**
 * Height of header element
 */
const HEADER_HEIGHT = 56 + 30 + 12 + 1; // header + subheader + margin-bottom  + border

/**
 * Height of table item
 */
const TABLE_ITEM_HEIGHT = 32;

/**
 * Builds statistics information object for usage in template.
 * @param {object} statistics Statistics entity
 * @returns {object[]}
 */
const parseDataInfo = function(statistics) {
   var dataInfo = [];
   var lastEvents = [];
   var allScores = [];
   var homeTeam = statistics.homeParticipant.participantName;
   var awayTeam = statistics.awayParticipant.participantName;

   statistics.lastEvents.forEach(function (lastEvent) {
      // Creating the item with the relevant data for each event
      var item = {
         start: lastEvent.start,
         homeParticipant: lastEvent.homeParticipant.participantName,
         awayParticipant: lastEvent.awayParticipant.participantName,
         homeScore: lastEvent.scores[0].homeScore,
         awayScore: lastEvent.scores[0].awayScore,
         type: lastEvent.scores[0].type
      };

      // Matching the scores of previous events with the current home team
      if (homeTeam === item.homeParticipant) {
         item.scoreOnLeft = item.homeScore;
         item.scoreOnRight = item.awayScore;
      } else {
         item.scoreOnLeft = item.awayScore;
         item.scoreOnRight = item.homeScore;
      }

      for (var j = 0; j < lastEvent.scores.length; j++) {
         if (lastEvent.scores[j].hasOwnProperty('homeScore') && lastEvent.scores[j].hasOwnProperty('awayScore')) {
            allScores.push(lastEvent.scores[j].homeScore, lastEvent.scores[j].awayScore);
         }
      }

      if (lastEvent.scores.length && lastEvent.scores[0].hasOwnProperty('homeScore') && lastEvent.scores[0].hasOwnProperty('awayScore')) {
         lastEvents.push(item);
      }
   });

   dataInfo.push({
      eventId: statistics.eventId,
      homeTeam: homeTeam,
      awayTeam: awayTeam,
      lastEvents: lastEvents,
      maxScore: Math.max.apply(null, allScores)
   });

   return dataInfo;
};

class HeadToHeadWidget extends React.Component {

   /**
    * Constructor
    * @param {object} props Widget attributes
    */
   constructor(props) {
      super(props);

      widgetModule.enableWidgetTransition(true);

      this.state = {
         events: [],
         stats: []
      };
   }

   /**
    * Called before mounting the widget.
    */
   componentWillMount() {
      this.refresh(this.props.args.eventId);
   }

   /**
    * Called on attributes change.
    * @param {object} nextProps New attributes
    */
   componentWillReceiveProps(nextProps) {
      this.refresh(nextProps.args.eventId);
   }

   /**
    * Fetches external data and updates widget.
    * @param {number} eventId Event identifier
    */
   refresh(eventId) {
      if (!eventId) {
         console.warn('eventId set from pageParam');
         eventId = coreLibrary.pageInfo.pageParam;
      } else {
         console.warn('eventId set from args.eventId');
      }

      statisticsModule.getHeadToHeadStatistics(eventId)
         .then((statistics) => {
            this.setState({
               events: parseDataInfo(statistics),
               stats: statistics.lastEvents
            });

            document.getElementsByTagName('body')[0].style.display = 'block';

            widgetModule.setWidgetHeight(
               this.state.stats.reduce(contentHeight => contentHeight + TABLE_ITEM_HEIGHT, HEADER_HEIGHT)
            );
         })
         .catch((error) => {
            console.warn(`Cannot load statistics data: ${error.message}`);
            widgetModule.removeWidget();
         });
   }

   /**
    * Creates widget template.
    * @returns {XML}
    */
   render() {
      const t = translationModule.getTranslation.bind(translationModule);

      return (
         <div className="KambiWidget-card-background-color KambiWidget-card-text-color KambiWidget-card-border-color l-flexbox l-vertical l-expander">
            <header className="KambiWidget-font kw-header l-flexbox l-align-center l-pt-16 l-pb-16 l-pl-16">{t('Head to head')}</header>
            {this.state.events.map((event) => {
               return (
                  <main key={event.eventId} className="KambiWidget-font l-flexbox l-vertical l-mb-12">
                     <div className="kw-teams l-flexbox l-flex-1 l-pl-16 l-pr-16 l-mb-6">
                        <div data-item-attr="home" className="l-flexbox l-flex-1 l-pack-start l-align-center l-pr-6">
                           <div className="text-truncate">{t(event.homeTeam)}</div>
                        </div>
                        <div data-item-attr="away" className="l-flexbox l-flex-1 l-pack-end l-align-center">
                           <div className="text-truncate">{t(event.awayTeam)}</div>
                        </div>
                     </div>
                     {event.lastEvents.map((score) => {
                        return (
                           <div
                              className="kw-table-item KambiWidget-card-background-color--hoverable l-flexbox l-flex-1 l-pl-16 l-pr-16"
                              key={score.start}
                           >
                              <ScoreBar side="home" ourScore={score.scoreOnLeft} theirScore={score.scoreOnRight} maxScore={event.maxScore} />
                              <ScoreBar side="away" ourScore={score.scoreOnRight} theirScore={score.scoreOnLeft} maxScore={event.maxScore} />
                           </div>
                        );
                     })}
                  </main>
               );
            })}
         </div>
      );
   }

}

HeadToHeadWidget.propTypes = {
   /**
    * Widget arguments
    */
   args: React.PropTypes.object.isRequired
};

export default HeadToHeadWidget;
