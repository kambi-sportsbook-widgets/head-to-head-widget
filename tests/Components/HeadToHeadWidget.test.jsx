/* eslint-env jest */
import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import { mount } from 'enzyme';
import HeadToHeadWidget from '../../src/js/Components/HeadToHeadWidget';
import { widgetModule } from 'kambi-widget-core-library';

let renderer;

jest.mock('kambi-widget-core-library', () => ({
   widgetModule: {
      adaptWidgetHeight: jest.fn()
   },
   translationModule: {
      getTranslation: jest.fn((key) => `Translated ${key}`)
   }
}));

const mockEvent0 = {
   homeParticipant: 'Poland',
   homeScore: 3,
   awayParticipant: 'Sweden',
   awayScore: 0,
   start: new Date('2017-05-05').toISOString()
};

const mockEvent1 = {
   homeParticipant: 'Montenegro',
   homeScore: 1,
   awayParticipant: 'Serbia',
   awayScore: 0,
   start: new Date('2017-05-04').toISOString()
};

describe('HeadToHeadWidget.render()', () => {

   beforeEach(() => {
      renderer = new ReactShallowRenderer();
      widgetModule.adaptWidgetHeight.mockClear();
   });

   it('renders correctly', () => {
      expect(renderer.render(
         <HeadToHeadWidget events={[{lastEvents: [mockEvent0, mockEvent1]}]} />
      )).toMatchSnapshot();
   });

});

describe('HeadToHeadWidget lifecycle methods', () => {

   beforeEach(() => {
      widgetModule.adaptWidgetHeight.mockClear();
   });

   it('initializes correctly', () => {
      expect(widgetModule.adaptWidgetHeight).not.toHaveBeenCalled();

      const wrapper = mount(
         <HeadToHeadWidget events={[]} />
      );

      expect(widgetModule.adaptWidgetHeight).toHaveBeenCalledTimes(1);
   });

   it('updates correctly', () => {
      const wrapper = mount(
         <HeadToHeadWidget events={[]} />
      );

      widgetModule.adaptWidgetHeight.mockClear();

      wrapper.setProps({events: []});

      expect(widgetModule.adaptWidgetHeight).toHaveBeenCalledTimes(1);
   });

});
