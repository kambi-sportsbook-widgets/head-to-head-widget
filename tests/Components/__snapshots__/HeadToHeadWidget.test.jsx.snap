// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeadToHeadWidget.render() renders correctly 1`] = `
<div>
  <Header
    title="Translated Head to head"
  />
  <BoxContainer>
    <EventInfo
      awayName="Sweden"
      awayScore={0}
      homeName="Poland"
      homeScore={3}
      start={1493942400000}
    />
  </BoxContainer>
  <BoxContainer>
    <EventInfo
      awayName="Serbia"
      awayScore={0}
      homeName="Montenegro"
      homeScore={1}
      start={1493856000000}
    />
  </BoxContainer>
</div>
`;
